### PR TITLE
Make ValueObjectPrinter's handling of its ValueObject pointers more principled (NFC)

### PR DESCRIPTION
**Pull request by jimingham**
I get a small but fairly steady stream of crash reports which I can only explain by ValueObjectPrinter trying to access its m_valobj field, and finding it NULL.  I have never been able to reproduce any of these, and the reports show a state too long after the fact to know what went wrong.

I've read through this section of lldb a bunch of times trying to figure out how this could happen, but haven't ever found anything actually wrong that could cause this.  OTOH, ValueObjectPrinter is somewhat sloppy about how it handles the ValueObject it is printing.

a) lldb allows you to make a ValueObjectPrinter with a Null incoming ValueObject.  However, there's no affordance to set the ValueObject in the Printer after the fact, and it doesn't really make sense to do that.  So I change the ValueObjectPrinter API's to take a ValueObject reference, rather than a pointer.  All the places that make ValueObjectPrinters already check the non-null status of their ValueObject's before making the ValueObjectPrinter, so sadly, I didn't find the bug, but this will enforce the intent.
b) The next step in printing the ValueObject is deciding which of the associated DynamicValue/SyntheticValue we are actually printing (based on the use_dynamic and use_synthetic settings in the original ValueObject.  This was put in a pointer by GetMostSpecializedValue, but most of the printer code just accessed the pointer, and it was hard to reason out whether we were guaranteed to always call this before using m_valobj.  So far as I could see we always do (sigh, didn't find the bug there either) but this was way too hard to reason about.

In fact, we figure out once which ValueObject we're going to print and don't change that through the life of the printer.  So I changed this to both set the "most specialized value" in the constructor, and then to always access it through GetMostSpecializedValue().  That makes it easier to reason about the use of this ValueObject as well.

This is an NFC change, all it does is make the code easier to reason about.

### CODE REVIEW

File: lldb/source/Commands/CommandObjectFrame.cpp [hunk 1]
```diff
@@ -177,7 +177,10 @@ class CommandObjectFrameDiagnose : public CommandObjectParsed {
 
     DumpValueObjectOptions options;
     options.SetDeclPrintingHelper(helper);
-    ValueObjectPrinter printer(valobj_sp.get(), &result.GetOutputStream(),
+    // We've already handled the case where the value object sp is null, so
+    // this is just to make sure future changes don't skip that:
+    assert(valobj_sp.get() && "Must have a valid ValueObject to print");
```
Comment (clayborg) [inline lldb/source/Commands/CommandObjectFrame.cpp:182]: It is fine to assert here, but we should not crash if valobj_sp is NULL. I would rather log or report an error that crash if possible.

File: lldb/source/Commands/CommandObjectFrame.cpp [hunk 2]
```diff
@@ -177,7 +177,10 @@ class CommandObjectFrameDiagnose : public CommandObjectParsed {
...
+    ValueObjectPrinter printer(*(valobj_sp.get()), &result.GetOutputStream(),
```
Comment (clayborg) [inline lldb/source/Commands/CommandObjectFrame.cpp:183]: Can this just be:
```
ValueObjectPrinter printer(*valobj_sp, &result.GetOutputStream(),
```

File: lldb/source/DataFormatters/ValueObjectPrinter.cpp [hunk 3]
```diff
@@ -18,39 +18,35 @@
...
 
-ValueObjectPrinter::ValueObjectPrinter(ValueObject *valobj, Stream *s) {
-  if (valobj) {
-    DumpValueObjectOptions options(*valobj);
-    Init(valobj, s, options, m_options.m_max_ptr_depth, 0, nullptr);
-  } else {
-    DumpValueObjectOptions options;
-    Init(valobj, s, options, m_options.m_max_ptr_depth, 0, nullptr);
-  }
+ValueObjectPrinter::ValueObjectPrinter(ValueObject &valobj, Stream *s)
```
Comment (clayborg) [inline lldb/source/DataFormatters/ValueObjectPrinter.cpp:21]: Would it make sense to pass in a "ValueObjectSP" here instead? Could some of our crashes be due to someone creating a ValueObjectPrinter and then somehow the ValueObject gets freed behind out backs?

**Review by clayborg — COMMENTED** [review]
See inlined comments, but maybe we need to start storing a `std::weak_ptr<ValueObject>` instead of raw pointers to things, and possibly always passsing a `ValueObjectSP` object into the `ValueObjectPrinter` so it holds a strong reference to the value object while printing. I could see dead pointers causing problems and crashes.

It would be nice to also not crash, right now we do things like:
```
assert(valobj);
valobj->CrashNow();
```
If we use `std::weak_ptr<ValueObject>` instead of raw pointers or references, we can then call the `bool std::weak_ptr<T>::expired() const` function to see if the weak pointer was valid before, but the main shared pointer keeping it alive has been freed. This will allow us to detect issues where our backing object has been freed.

**Comment by jimingham** [comment]
Replying to a bunch of Greg's comments in sum.  

The way we use the ValueObjectPrinter is that you have a ValueObject and want to print it, so you make a ValueObjectPrinter passing in that ValueObject, then tell it to print it.  We don't preserve ValueObjectPrinters or try to reuse them.  There's no attempt to track changes in the ValueObject, for instance if you changed the use dynamic setting of the underlying ValueObject that won't effect the ValueObjectPrinter you made before you changed the setting.  It doesn't track the SyntheticChildProviderGeneration, so if you changed synthetic child providers on that ValueObject, it wouldn't track that, etc.

I think that's a useful simplification.   We don't have to reason about lifecycles in the ValueObjectPrinter.   The caller is responsible for the lifecycle of the ValueObject being printed - and that ValueObject's ClusterManager is responsible for all the children & synthetic children of that ValueObject.  The ValueObjectPrinter gets a reference to it because it expects to do some work on the ValueObject and then get discarded, which is how we always use it.

I can put in a comment to this effect, though the fact that we're taking in a `ValueObject &` and not a shared pointer also implies this is the way you are supposed to use it.

File: lldb/source/Commands/CommandObjectFrame.cpp (hunk 1 above)
Comment (jimingham) [reply lldb/source/Commands/CommandObjectFrame.cpp:182]: I'm of two minds about this.  The current code does:

   if (!valobj_sp) {
      // do some error handling
      return;
  }
  // enough code to push this check off the top of your monitor screen

Then this code.  So I could have just duplicated the check & error handling again before printing the thing, but that looked bogus to me.  OTOH, I don't want someone to forget check, so I put in the assert, so if one of us accidentally deletes the check, we'll know.

But if that bugs you I'm happy to cut and paste the error handling here again.  It just looked weird to me.

File: lldb/source/Commands/CommandObjectFrame.cpp (hunk 2 above)
Comment (jimingham) [reply lldb/source/Commands/CommandObjectFrame.cpp]: Apparently it can!  cool...

File: lldb/source/DataFormatters/ValueObjectPrinter.cpp (hunk 3 above)
Comment (jimingham) [reply lldb/source/DataFormatters/ValueObjectPrinter.cpp:21]: For this series of crashes,  the EXC_BAD_ACCESS address is the correct offset for the ivar being accessed.  So the address of the ValueObject pointer has to be 0x0.  If the pointer had gotten freed, the printer's pointer value wouldn't change, so the access would just be some random looking address.

Also, see the more general comment about uses of the ValueObjectPrinter.

**Comment by jimingham** [comment]
Note, these really are transitory objects and need to be quick to make and delete because in the course of printing a complex object we make a ValueObjectPrinter for each node we print.  So we're making a lot of them as we descend and then directly throwing them away when we're done.  So it makes sense to treat them as transitory in all their uses, IMO.

**Comment by clayborg** [comment]
None of what I suggested was mandatory. I was mainly trying to make sure things don't go away on us without us knowing about it. If we have crashes it is always hard to tell if simple heap corruption is the issue or if we are having lifetime issues. If you are not concerned about lifetime issues then you can ignore the `std::weak_ptr<T>` suggestion, but if you are, this would be a good way to tell as we can check `std::weak_ptr<T>::expired()`. 

Running a lldb with libgmalloc is a good way to watch for pointer lifetime issues and things being freed if someone ever has a repro case that they add to a bug.

**Comment by jimingham** [comment]
I don't think this is a lifetime issue.  If it were I would see more of the crashes at random addresses.

Whatever's going on is something odd, so I'm just trying to make the code easier to reason about.  I think in the ValueObjectPrinter's case, it's code will be cleaner if it expects to be used as it is currently used, as a one shot printer that is not responsible for the lifecycle of the ValueObject it's printing.  That was the original intent; I'm not really changing that.  I just got frustrated while trying to find the bug by inspection.  I kept going "ooh, there's an unchecked use of m_valobj, I wonder if that could cause problems?" only to follow all the paths through the code and end up convincing myself that you couldn't get here with a NULL pointer, so that wasn't the problem.  If we make it so internally to ValueObjectPrinter you can't get your hands on something that might be null, and I see another of these crashes, I'll be able to just look outside ValueObjectPrinter for the answer.

Jim

**Comment by jimingham** [comment]
The other thing I'm trying to eliminate is problem arising from the way the old code started with `m_valobj` primed to `nullptr` and then getting set by calling GetMostSpecializedValue, but then not always accessing it through that API, but often accessing `m_valobj` directly.  So you could by accident have forgotten to call GetMostSpecializedValue before you accessed the `m_valobj`.  Again, I read closely through the code to see if this could happen IRL and I don't see how it could.  By making that impossible, if that was the problem, this should define it away.

**Review by clayborg — APPROVED** [review]
Ok, LGTM.

**Pull request merged**